### PR TITLE
First version of SQL Server provider

### PR DESCRIPTION
**Pull request by msclash**
we want to add SQL Server Monitoring to the Azure Monitoring Solution and this PR contains the provider for SQL Server

### CODE REVIEW

File: sapmon/payload/provider/sqlserver.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                                     SQLHostname: str = None,
+                                     SQLUser: str = None,
+                                     SQLPassword: str = None) -> pyodbc.Connection:
+      if not SQLHostname:
+         SQLHostname = self.SQLHostname
+      if not SQLUser:
+         SQLUser = self.SQLUser
+      if not SQLPassword:
+         SQLPassword = self.SQLPassword
+      self.tracer.debug("Connection  : DRIVER={ODBC Driver 17 for SQL Server};SERVER=%s;UID=%s;PWD=%s" % (SQLHostname, SQLUser, SQLPassword))
```
Comment (persiaAziz) [inline sapmon/payload/provider/sqlserver.py:119]: Does this driver version '17' work with any SQL server versions?

Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:119]: yes, it supports all SQL Server Versions that are supported on Azure.

File: sapmon/payload/provider/sqlserver.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            self.tracer.error("[%s] error accessing the separate KeyVault (%s)" % (self.fullName,e))
+            return False
+
+         # Access the actual secret from the external KeyVault
+         # todo: proper (provider-independent) handling of external KeyVaults
+         try:
+            self.SQLPassword = kv.getSecret(SQLPassword, None).value
+         except Exception as e:
+            self.tracer.error("[%s] error accessing the secret inside the separate KeyVault (%s)" % (self.fullName,e))
+            return False
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:82]: need to remove external KV until we have generic handling for this @PakDLiu

File: sapmon/payload/provider/sqlserver.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+      try:
+        connection = self.providerInstance._establishsqlConnectionToHost()
+        #Validate that we're indeed connected
+        #if connection.isconnected():
+      except Exception as e:
+         self.tracer.warning("[%s] could not connect to sql (%s) " % (self.fullName,e))
+         return (None)
+      return (connection)
+
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:149]: nit: remove blank line

File: sapmon/payload/provider/sqlserver.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+            self.tracer.error("[%s] unable to validate connection status" % self.fullName)
+            return False
+      except Exception as e:
+         self.tracer.error("[%s] could not establish sql connection %s (%s)" % (self.fullName,self.sqlHostname,e))
+         return False
+
+      # Try to run a query against the services view
+      # This query will (rightfully) fail if the sql license is expired
+      try:
+         # execute todo
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:102]: validation query is missing

File: sapmon/payload/provider/sqlserver.py [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+                                     SQLPassword: str = None) -> pyodbc.Connection:
+      if not SQLHostname:
+         SQLHostname = self.SQLHostname
+      if not SQLUser:
+         SQLUser = self.SQLUser
+      if not SQLPassword:
+         SQLPassword = self.SQLPassword
+      self.tracer.debug("Connection  : DRIVER={ODBC Driver 17 for SQL Server};SERVER=%s;UID=%s;PWD=%s" % (SQLHostname, SQLUser, SQLPassword))
+
+      conn = pyodbc.connect("DRIVER={ODBC Driver 17 for SQL Server};SERVER=%s;UID=%s;PWD=%s" % (SQLHostname, SQLUser, SQLPassword))
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:121]: refactor to use retryer @PakDLiu

File: sapmon/payload/provider/sqlserver.py [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+      if len(resultRows) == 0:
+         self.tracer.debug("[%s] result set is empty" % self.fullName)
+         return None
+      resultHash = None
+      try:
+         resultHash = hashlib.md5(str(resultRows).encode("utf-8")).hexdigest()
+         self.tracer.debug("resultHash=%s" % resultHash)
+      except Exception as e:
+         self.tracer.error("[%s] could not calculate result hash (%s)" % (self.fullName,e))
+      return resultHash
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:164]: not sure if this is needed here

File: sapmon/payload/provider/sqlserver.py [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+         return value.decode('utf-16le')
+      self.tracer.info("[%s] connecting to sql and executing SQL" % self.fullName)
+
+      # Find and connect to sql server
+      connection = self._getsqlConnection()
+      if not connection:
+         return False
+
+      cursor = connection.cursor()
+      connection.add_output_converter(-150, handle_sql_variant_as_string)
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:213]: what does the -150 do?

File: sapmon/payload/provider/sqlserver.py [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+
+      # Execute SQL statement
+      try:
+         self.tracer.debug("[%s] executing SQL statement %s" % (self.fullName, sql))
+         cursor.execute(sql)
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:218]: refactor to use retryer @PakDLiu

File: sapmon/payload/provider/sqlserver.py [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+         self.tracer.debug("[%s] closing sql connection" % self.fullName)
+         connection.close()
+      except Exception as e:
+         self.tracer.error("[%s] could not close connection to sql instance (%s)" % (self.fullName,e))
+         return False
+
+      self.tracer.info("[%s] successfully ran SQL for check" % self.fullName)
+      return True
+
+
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:246]: nit: remove blank line

File: sapmon/payload/provider/sqlserver.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+from helper.azure import *
+from helper.tools import *
+from provider.base import ProviderInstance, ProviderCheck
+from typing import Dict, List
+
+###############################################################################
+
+
+###############################################################################
+
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:21]: nit: remove lines

File: sapmon/payload/provider/sqlserver.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
+# Python modules
+import hashlib
+import json
+import logging
+import re
+import time
+import pyodbc
+
+
```
Comment (tniek) [inline sapmon/payload/provider/sqlserver.py:9]: nit: remove blank line

File: sapmon/payload/provider/sqlserver.py (hunk 8 above)
Comment (PakDLiu) [reply sapmon/payload/provider/sqlserver.py:218]: The method `_actionExecuteSql` should return `None`, and the retry logic will retry if it catches any exception

File: sapmon/payload/provider/sqlserver.py (hunk 7 above)
Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:213]: I took this over from the HANA provider

File: sapmon/payload/provider/sqlserver.py (hunk 3 above)
Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:149]: Fixed

File: sapmon/payload/provider/sqlserver.py (hunk 6 above)
Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:164]: it is not used in the provider itself and if it is not called from the outside we can remove it

File: sapmon/payload/provider/sqlserver.py (hunk 8 above)
Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:218]: the method returns either True or False (same as the Hana provider), will this trigger the retry logic as well (in the case of "false" of course)

File: sapmon/payload/provider/sqlserver.py (hunk 4 above)
Comment (msclash) [reply sapmon/payload/provider/sqlserver.py:102]: fixed

**Comment by PakDLiu** [comment]
Hey Clas, I have updated your branch to be able to use the retryer.
Is there an environment where I can test the change? Or can you test the changes?
Thanks

File: sapmon/payload/provider/sqlserver.py (hunk 5 above)
Comment (tniek) [reply sapmon/payload/provider/sqlserver.py:101]: will make a separate change for both DB providers

**Pull request merged**